From 79f0312b25a422879e44d5a79205343e5b8c5e44 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tomasz=20Paku=C5=82a?= <tomasz.pakula.oficjalny@gmail.com>
Date: Thu, 15 Jan 2026 19:28:20 +0100
Subject: [PATCH 11/11] drm/amd/display: HDMI ALLM in HDMI Forum vsif support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Tells sinks to enable gaming mode automatically which should be always
used with PC as a source.

ID based on public HDMI Forum VSDB IOU, rest based on previous vsdb and
some time with trying things out randomly.

Signed-off-by: Tomasz Pakuła <tomasz.pakula.oficjalny@gmail.com>
---
 .../gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c |  8 ++-
 drivers/gpu/drm/amd/display/dc/core/dc.c      |  5 ++
 .../gpu/drm/amd/display/dc/core/dc_resource.c | 21 ++++++-
 drivers/gpu/drm/amd/display/dc/dc_stream.h    |  2 +
 .../amd/display/dc/inc/hw/stream_encoder.h    |  1 +
 .../amd/display/modules/inc/mod_info_packet.h |  3 +
 .../display/modules/info_packet/info_packet.c | 62 +++++++++++++++----
 7 files changed, 86 insertions(+), 16 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
index 9ac0441d393b..4fb58adbc1d4 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
@@ -7264,8 +7264,10 @@ create_stream_for_sink(struct drm_connector *connector,
 
 	update_stream_signal(stream, sink);
 
-	if (stream->signal == SIGNAL_TYPE_HDMI_TYPE_A)
-		mod_build_hf_vsif_infopacket(stream, &stream->vsp_infopacket);
+	if (stream->signal == SIGNAL_TYPE_HDMI_TYPE_A) {
+		mod_build_hf_vsif_infopacket(stream, &stream->hfvsif_infopacket);
+		mod_build_forum_vsif_infopacket(&connector->display_info, &stream->forum_vsif_infopacket);
+	}
 
 	if (stream->signal == SIGNAL_TYPE_DISPLAY_PORT ||
 	    stream->signal == SIGNAL_TYPE_DISPLAY_PORT_MST ||
@@ -7291,6 +7293,8 @@ create_stream_for_sink(struct drm_connector *connector,
 		mod_build_vsc_infopacket(stream, &stream->vsc_infopacket, stream->output_color_space, tf);
 		aconnector->sr_skip_count = AMDGPU_DM_PSR_ENTRY_DELAY;
 
+		if (stream->link->dpcd_caps.dongle_type == DISPLAY_DONGLE_DP_HDMI_CONVERTER)
+			mod_build_forum_vsif_infopacket(&connector->display_info, &stream->forum_vsif_infopacket);
 	}
 finish:
 	dc_sink_release(sink);
diff --git a/drivers/gpu/drm/amd/display/dc/core/dc.c b/drivers/gpu/drm/amd/display/dc/core/dc.c
index 5f2d5638c819..b6ee6baa93e1 100644
--- a/drivers/gpu/drm/amd/display/dc/core/dc.c
+++ b/drivers/gpu/drm/amd/display/dc/core/dc.c
@@ -3298,6 +3298,9 @@ static void copy_stream_update_to_stream(struct dc *dc,
 	if (update->hfvsif_infopacket)
 		stream->hfvsif_infopacket = *update->hfvsif_infopacket;
 
+	if (update->forum_vsif_infopacket)
+		stream->forum_vsif_infopacket = *update->forum_vsif_infopacket;
+
 	if (update->vtem_infopacket)
 		stream->vtem_infopacket = *update->vtem_infopacket;
 
@@ -3606,6 +3609,7 @@ static void commit_planes_do_stream_update(struct dc *dc,
 					stream_update->vsc_infopacket ||
 					stream_update->vsp_infopacket ||
 					stream_update->hfvsif_infopacket ||
+					stream_update->forum_vsif_infopacket ||
 					stream_update->adaptive_sync_infopacket ||
 					stream_update->vtem_infopacket ||
 					stream_update->avi_infopacket) {
@@ -5078,6 +5082,7 @@ static bool full_update_required(struct dc *dc,
 			stream_update->vsc_infopacket ||
 			stream_update->vsp_infopacket ||
 			stream_update->hfvsif_infopacket ||
+			stream_update->forum_vsif_infopacket ||
 			stream_update->vtem_infopacket ||
 			stream_update->adaptive_sync_infopacket ||
 			stream_update->avi_infopacket ||
diff --git a/drivers/gpu/drm/amd/display/dc/core/dc_resource.c b/drivers/gpu/drm/amd/display/dc/core/dc_resource.c
index bc5dedf5f60c..b3ebf9435fff 100644
--- a/drivers/gpu/drm/amd/display/dc/core/dc_resource.c
+++ b/drivers/gpu/drm/amd/display/dc/core/dc_resource.c
@@ -4712,6 +4712,15 @@ static void set_hfvs_info_packet(
 
 	*info_packet = stream->hfvsif_infopacket;
 }
+static void set_forum_info_packet(
+		struct dc_info_packet *info_packet,
+		struct dc_stream_state *stream)
+{
+	if (!stream->forum_vsif_infopacket.valid)
+		return;
+
+	*info_packet = stream->forum_vsif_infopacket;
+}
 
 static void adaptive_sync_override_dp_info_packets_sdp_line_num(
 		const struct dc_crtc_timing *timing,
@@ -4796,6 +4805,7 @@ void resource_build_info_frame(struct pipe_ctx *pipe_ctx)
 	info->hdrsmd.valid = false;
 	info->vsc.valid = false;
 	info->hfvsif.valid = false;
+	info->forum_vsif.valid = false;
 	info->vtem.valid = false;
 	info->adaptive_sync.valid = false;
 	signal = pipe_ctx->stream->signal;
@@ -4809,6 +4819,7 @@ void resource_build_info_frame(struct pipe_ctx *pipe_ctx)
 
 		set_vendor_info_packet(&info->vendor, pipe_ctx->stream);
 		set_hfvs_info_packet(&info->hfvsif, pipe_ctx->stream);
+		set_forum_info_packet(&info->hfvsif, pipe_ctx->stream);
 		set_vtem_info_packet(&info->vtem, pipe_ctx->stream);
 
 		set_spd_info_packet(&info->spd, pipe_ctx->stream);
@@ -4822,9 +4833,13 @@ void resource_build_info_frame(struct pipe_ctx *pipe_ctx)
 
 		set_hdr_static_info_packet(&info->hdrsmd, pipe_ctx->stream);
 		set_adaptive_sync_info_packet(&info->adaptive_sync,
-										pipe_ctx->stream,
-										info,
-										vstartup_start);
+					      pipe_ctx->stream, info,
+					      vstartup_start);
+
+		/* Trigger ALLM if using DP->HDMI PCON */
+		if (pipe_ctx->stream->link->dpcd_caps.dongle_type ==
+		    DISPLAY_DONGLE_DP_HDMI_CONVERTER)
+			set_forum_info_packet(&info->hfvsif, pipe_ctx->stream);
 	}
 
 	patch_gamut_packet_checksum(&info->gamut);
diff --git a/drivers/gpu/drm/amd/display/dc/dc_stream.h b/drivers/gpu/drm/amd/display/dc/dc_stream.h
index 76cf9fdedab0..c9ae8ae77586 100644
--- a/drivers/gpu/drm/amd/display/dc/dc_stream.h
+++ b/drivers/gpu/drm/amd/display/dc/dc_stream.h
@@ -201,6 +201,7 @@ struct dc_stream_state {
 	struct dc_info_packet vsc_infopacket;
 	struct dc_info_packet vsp_infopacket;
 	struct dc_info_packet hfvsif_infopacket;
+	struct dc_info_packet forum_vsif_infopacket;
 	struct dc_info_packet vtem_infopacket;
 	struct dc_info_packet adaptive_sync_infopacket;
 	struct dc_info_packet avi_infopacket;
@@ -334,6 +335,7 @@ struct dc_stream_update {
 	struct dc_info_packet *vsc_infopacket;
 	struct dc_info_packet *vsp_infopacket;
 	struct dc_info_packet *hfvsif_infopacket;
+	struct dc_info_packet *forum_vsif_infopacket;
 	struct dc_info_packet *vtem_infopacket;
 	struct dc_info_packet *adaptive_sync_infopacket;
 	struct dc_info_packet *avi_infopacket;
diff --git a/drivers/gpu/drm/amd/display/dc/inc/hw/stream_encoder.h b/drivers/gpu/drm/amd/display/dc/inc/hw/stream_encoder.h
index 27f950ae45ee..0e6a1d2eea5d 100644
--- a/drivers/gpu/drm/amd/display/dc/inc/hw/stream_encoder.h
+++ b/drivers/gpu/drm/amd/display/dc/inc/hw/stream_encoder.h
@@ -83,6 +83,7 @@ struct encoder_info_frame {
 	struct dc_info_packet gamut;
 	struct dc_info_packet vendor;
 	struct dc_info_packet hfvsif;
+	struct dc_info_packet forum_vsif;
 	struct dc_info_packet vtem;
 	/* source product description */
 	struct dc_info_packet spd;
diff --git a/drivers/gpu/drm/amd/display/modules/inc/mod_info_packet.h b/drivers/gpu/drm/amd/display/modules/inc/mod_info_packet.h
index 91fe039c0c95..ad7cc6a9cd53 100644
--- a/drivers/gpu/drm/amd/display/modules/inc/mod_info_packet.h
+++ b/drivers/gpu/drm/amd/display/modules/inc/mod_info_packet.h
@@ -41,6 +41,9 @@ void mod_build_vsc_infopacket(const struct dc_stream_state *stream,
 void mod_build_hf_vsif_infopacket(const struct dc_stream_state *stream,
 		struct dc_info_packet *info_packet);
 
+void mod_build_forum_vsif_infopacket(const struct drm_display_info *display,
+		struct dc_info_packet *info_packet);
+
 enum adaptive_sync_type {
 	ADAPTIVE_SYNC_TYPE_NONE                  = 0,
 	ADAPTIVE_SYNC_TYPE_DP                    = 1,
diff --git a/drivers/gpu/drm/amd/display/modules/info_packet/info_packet.c b/drivers/gpu/drm/amd/display/modules/info_packet/info_packet.c
index 429bc0ab86d6..e1e9e019f874 100644
--- a/drivers/gpu/drm/amd/display/modules/info_packet/info_packet.c
+++ b/drivers/gpu/drm/amd/display/modules/info_packet/info_packet.c
@@ -47,6 +47,11 @@ enum vsc_packet_revision {
 #define HDMI_INFOFRAME_TYPE_VENDOR 0x81
 #define HF_VSIF_VERSION 1
 
+// HDMI 2.1
+#define HDMI_FORUM_VSIF_VERSION	 1
+#define HDMI_FORUM_VSIF_LENGTH	 27
+#define HDMI_FORUM_VSIF_ALLM_BIT 1
+
 // VTEM Byte Offset
 #define VTEM_PB0		0
 #define VTEM_PB1		1
@@ -441,6 +446,21 @@ void mod_build_vsc_infopacket(const struct dc_stream_state *stream,
 	}
 }
 
+static void set_hdmi_vsif_checksum(struct dc_info_packet *info_packet)
+{
+	uint8_t checksum = 0;
+	uint8_t length = info_packet->hb2 & 0x1F;
+
+	checksum += info_packet->hb0;
+	checksum += info_packet->hb1;
+	checksum += info_packet->hb2;
+
+	for (uint32_t i = 1; i <= length; i++)
+		checksum += info_packet->sb[i];
+
+	info_packet->sb[0] = (uint8_t) (0x100 - checksum);
+}
+
 /**
  *  mod_build_hf_vsif_infopacket - Prepare HDMI Vendor Specific info frame.
  *                                 Follows HDMI Spec to build up Vendor Specific info frame
@@ -453,8 +473,6 @@ void mod_build_hf_vsif_infopacket(const struct dc_stream_state *stream,
 {
 		unsigned int length = 5;
 		bool hdmi_vic_mode = false;
-		uint8_t checksum = 0;
-		uint32_t i = 0;
 		enum dc_timing_3d_format format;
 
 		info_packet->valid = false;
@@ -509,18 +527,40 @@ void mod_build_hf_vsif_infopacket(const struct dc_stream_state *stream,
 		info_packet->hb1 = 0x01;
 		info_packet->hb2 = (uint8_t) (length);
 
-		checksum += info_packet->hb0;
-		checksum += info_packet->hb1;
-		checksum += info_packet->hb2;
-
-		for (i = 1; i <= length; i++)
-			checksum += info_packet->sb[i];
-
-		info_packet->sb[0] = (uint8_t) (0x100 - checksum);
-
+		set_hdmi_vsif_checksum(info_packet);
 		info_packet->valid = true;
 }
 
+/**
+ *  mod_build_forum_vsif_infopacket - Prepare HDMI Forum Vendor Specific info frame.
+ *
+ *  @stream:      contains data we may need to construct VSIF
+ *  @info_packet: output structure where to store VSIF
+ */
+void mod_build_forum_vsif_infopacket(const struct drm_display_info *info,
+		struct dc_info_packet *info_packet)
+{
+	info_packet->valid = false;
+
+	/* Not needed besides ALLM */
+	if (!info->hdmi.allm)
+		return;
+
+	info_packet->hb0 = HDMI_INFOFRAME_TYPE_VENDOR;
+	info_packet->hb1 = 0x01;
+	info_packet->hb2 = HDMI_FORUM_VSIF_LENGTH;
+
+	info_packet->sb[1] = HDMI_FORUM_IEEE_OUI & 0xff;
+	info_packet->sb[2] = (HDMI_FORUM_IEEE_OUI >> 8) & 0xff;
+	info_packet->sb[3] = (HDMI_FORUM_IEEE_OUI >> 16) & 0xff;
+	info_packet->sb[4] = HDMI_FORUM_VSIF_VERSION;
+
+	info_packet->sb[5] = (uint8_t) (info->hdmi.allm << HDMI_FORUM_VSIF_ALLM_BIT);
+
+	set_hdmi_vsif_checksum(info_packet);
+	info_packet->valid = true;
+}
+
 void mod_build_adaptive_sync_infopacket(const struct dc_stream_state *stream,
 		enum adaptive_sync_type asType,
 		const struct AS_Df_params *param,
-- 
2.52.0

